modules/aws/ssm.go

```diff
@@ -68,3 +75,141 @@ func NewSsmClientE(t testing.TestingT, region string) (*ssm.SSM, error) {
+		}
+
+		return "", nil
+	})
+
+	return err
+}
+
+// WaitForSsmInstance waits until the instance get registered to the SSM inventory.
+func WaitForSsmInstance(t testing.TestingT, awsRegion, instanceID string, timeout time.Duration) {
+	if err := WaitForSsmInstanceE(t, awsRegion, instanceID, timeout); err != nil {
+		t.Fatalf("failed to find %s in the SSM store: %s", instanceID, err.Error())
+	}
+}
+
+// CheckSsmCommand checks that you can run the given command on the given instance through AWS SSM. Returns stdout and stderr.
+func CheckSsmCommand(t testing.TestingT, awsRegion, instanceID, command string, timeout time.Duration) (string, string) {
+	stdout, stderr, err := CheckSsmCommandE(t, awsRegion, instanceID, command, timeout)
+	if err != nil {
+		t.Fatalf("failed to execute '%s' on %s (%v):]\n  stdout: %#v\n  stderr: %#v", command, instanceID, err, stdout, stderr)
+	}
+	return stdout, stderr
+}
+
+type result struct {
+	Stdout string
+	Stderr string
+}
+
+// CheckSsmCommandE checks that you can run the given command on the given instance through AWS SSM. Returns the stdout, stderr and an error if one occurs.
+func CheckSsmCommandE(t testing.TestingT, awsRegion, instanceID, command string, timeout time.Duration) (string, string, error) {
+	logger.Logf(t, "Running command %s on %s", command, instanceID)
+
+	timeBetweenRetries := 2 * time.Second
+	maxRetries := int(timeout.Seconds() / timeBetweenRetries.Seconds())
+
+	// Now that we know the instance in the SSM inventory, we can send the command
+	client, err := NewSsmClientE(t, awsRegion)
+	if err != nil {
+		return "", "", err
+	}
+	comment := "Terratest SSM"
+	documentName := "AWS-RunShellScript"
+	req, resp := client.SendCommandRequest(&ssm.SendCommandInput{
+		Comment:      &comment,
+		DocumentName: &documentName,
+		InstanceIds:  []*string{&instanceID},
+		Parameters: map[string][]*string{
+			"commands": []*string{&command},
+		},
+	})
+	if err := req.Send(); err != nil {
+		return "", "", err
+	}
+
+	// Wait for the result
+	description := "Waiting for the result of the command"
+	retryableErrors := map[string]string{
+		"InvocationDoesNotExist": "InvocationDoesNotExist",
+		"bad status: Pending":    "bad status: Pending",
+		"bad status: InProgress": "bad status: InProgress",
+		"bad status: Delayed":    "bad status: Delayed",
+	}
+	out, err := retry.DoWithRetryableErrorsE(t, description, retryableErrors, maxRetries, timeBetweenRetries, func() (string, error) {
+		client, err := NewSsmClientE(t, awsRegion)
+		if err != nil {
+			return "", err
+		}
+		req, resp := client.GetCommandInvocationRequest(&ssm.GetCommandInvocationInput{
+			CommandId:  resp.Command.CommandId,
+			InstanceId: &instanceID,
+		})
+		if err := req.Send(); err != nil {
+			return "", err
+		}
+
+		// Remove the SSM prefix from stderr
+		stdErr := *resp.StandardErrorContent
+		prefix := regexp.MustCompile(`/var/lib/amazon/ssm/.*/_script\.sh: line 1: `)
```

Comment:
It would either never panic or always, wouldn't it?

Since we already know it's a valid regexp pattern it should be safe to keep that way right?

Reply:
We want to avoid panics anywhere we can, especially in Terratest, as even a single panic could prevent clean up and leave lots of expensive resources running. E.g., I wouldn't want to kick off our test suite with this code in there now due to the possibility of a panic (same for various nil pointer dereferences I mentioned elsewhere) and even if it works now, someone could make a harmless-looking change to it in the future that causes panics.

Reply:
I had written that because in some cases SSM adds a weird prefix when an error occurs but I can't get it to do it know. Maybe it only happens when we send multiple commands at once. Since I can't write a test case for this, let's remove it for now.

I think this was the last item of your review so everything should be good now 🎉